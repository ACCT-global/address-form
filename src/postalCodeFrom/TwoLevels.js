import React, { Component } from 'react'
import PropTypes from 'prop-types'
import AddressShape from '../propTypes/AddressShape'
import SelectLevel from './SelectLevel'
import SelectPostalCode from './SelectPostalCode'

class TwoLevels extends Component {
  render() {
    const { address, rules, onChangeAddress } = this.props

    return (
      <div>
        <SelectLevel
          level={0}
          rules={rules}
          address={address}
          onChangeAddress={onChangeAddress}
        />
        <SelectPostalCode
          address={address}
          rules={rules}
          onChangeAddress={onChangeAddress}
        />
      </div>
    )
  }
}

TwoLevels.propTypes = {
  address: PropTypes.shape(AddressShape),
  rules: PropTypes.object.isRequired,
  onChangeAddress: PropTypes.func.isRequired,
}

export default TwoLevels
