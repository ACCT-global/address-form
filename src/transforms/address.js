import reduce from 'lodash/reduce'
import filter from 'lodash/filter'
import map from 'lodash/map'
import difference from 'lodash/difference'
import find from 'lodash/find'
import findIndex from 'lodash/findIndex'
import { getField } from '../selectors/fields'
import { validateChangedFields } from '../validateAddress'
import msk from 'msk'

export function addValidation(address) {
  return reduce(
    address,
    (newAddress, value, propName) => {
      newAddress[propName] = { value }
      return newAddress
    },
    {}
  )
}

export function removeValidation(address) {
  return reduce(
    address,
    (newAddress, { value }, propName) => {
      newAddress[propName] = value
      return newAddress
    },
    {}
  )
}

export function addNewField(address, fieldName, value) {
  return reduce(
    address,
    (newAddress, prop, propName) => {
      newAddress[propName] = {
        ...prop,
        [fieldName]: value,
      }
      return newAddress
    },
    {}
  )
}

export function removeField(address, fieldName) {
  return reduce(
    address,
    (newAddress, prop, propName) => {
      newAddress[propName] = { ...prop }
      delete newAddress[propName][fieldName]
      return newAddress
    },
    {}
  )
}

export function addDisabledToProtectedFields(fields, rules) {
  return reduce(
    fields,
    (newFields, prop, propName) => {
      const hasValue = prop && prop.value
      const isProtectField =
        rules.postalCodeProtectedFields &&
        rules.postalCodeProtectedFields.indexOf(propName) !== -1

      newFields[propName] = prop

      if (isProtectField && hasValue) {
        newFields[propName] = {
          ...prop,
          disabled: true,
        }
      }

      return newFields
    },
    {}
  )
}

export function handleMultipleValues(fields) {
  return reduce(
    fields,
    (newFields, prop, propName) => {
      const hasMultipleValues = prop.value.indexOf(';') !== -1

      newFields[propName] = prop

      if (hasMultipleValues) {
        newFields[propName] = {
          ...prop,
          value: null,
          valueOptions: prop.value.split(';'),
        }
      }

      return newFields
    },
    {}
  )
}

export function maskFields(addressFields, rules) {
  return reduce(
    addressFields,
    (newAddressFields, prop, propName) => {
      const fieldRule = getField(propName, rules)

      newAddressFields[propName] = prop

      if (fieldRule && fieldRule.mask) {
        newAddressFields[propName] = {
          ...prop,
          ...(prop.value ? { value: msk(prop.value, fieldRule.mask) } : {}),
        }
      }

      return newAddressFields
    },
    {}
  )
}

export function addFocusToNextInvalidField(fields, rules) {
  const invalidFilledField = getFirstInvalidFilledField(fields, rules)

  if (invalidFilledField) {
    const { fieldName, field } = invalidFilledField

    return {
      ...fields,
      [fieldName]: field,
    }
  }

  const requiredField = getFirstRequiredFieldNotFilled(fields, rules)

  if (requiredField) {
    const { fieldName, field } = requiredField

    return {
      ...fields,
      [fieldName]: field,
    }
  }

  return fields
}

function getFirstInvalidFilledField(fields, rules) {
  const allFieldsVisited = addNewField(fields, 'visited', true)
  const validatedFields = validateChangedFields(
    allFieldsVisited,
    allFieldsVisited,
    rules
  )

  const orderedValidatedFields = orderFieldsByRules(validatedFields, rules)

  const firstInvalidField = find(
    orderedValidatedFields,
    field => field && field.valid === false
  )

  if (firstInvalidField) {
    return {
      fieldName: firstInvalidField.name,
      field: {
        ...validatedFields[firstInvalidField.name],
        focus: true,
      },
    }
  }

  return null
}

function orderFieldsByRules(fields, rules) {
  return reduce(
    fields,
    (acc, field, fieldName) => {
      const index = findIndex(
        rules.fields,
        ruleField => ruleField.name === fieldName
      )
      if (index === -1) return acc

      acc[index] = { ...field, name: fieldName }
      return acc
    },
    Array(Object.keys(fields).length)
  )
}

function getFirstRequiredFieldNotFilled(fields, rules) {
  const requiredFields = filter(rules.fields, field => field.required)
  const requiredFieldsNames = map(requiredFields, field => field.name)

  const fieldsNames = Object.keys(fields)
  const requiredFieldNotFilled = difference(requiredFieldsNames, fieldsNames)

  if (requiredFieldNotFilled && requiredFieldNotFilled.length > 0) {
    const nextRequiredFieldName = requiredFieldNotFilled[0]

    return { fieldName: nextRequiredFieldName, field: { focus: true } }
  }

  return null
}
