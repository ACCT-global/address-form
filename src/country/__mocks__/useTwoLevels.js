import { TWO_LEVELS } from '../../constants'
import { secondLevelPostalCodes } from '../../transforms/postalCodes'
import {
  getOneLevel,
  getTwoLevels,
} from '../../transforms/addressFieldsOptions'

const countryData = {
  'I Región': {
    Camiña: '1150000',
    Colchane: '1160000',
    Huara: '1140000',
    Iquique: '1100000',
    Pica: '1170000',
  },
  'II Región': {
    Antofagasta: '1240000',
    Calama: '1390000',
  },
}

export default {
  postalCodeFrom: TWO_LEVELS,
  postalCodeLevels: ['state', 'neighborhood'],
  secondLevelPostalCodes: secondLevelPostalCodes(countryData),
  fields: [
    {
      name: 'street',
      label: 'street',
      required: true,
      size: 'xlarge',
    },
    {
      name: 'number',
      label: 'number',
      required: true,
      size: 'mini',
    },
    {
      name: 'complement',
      label: 'complement',
      size: 'large',
    },
    {
      name: 'state',
      label: 'region',
      required: true,
      size: 'large',
      isUpperCase: false,
      options: getOneLevel(countryData),
    },
    {
      name: 'neighborhood',
      label: 'community',
      required: true,
      size: 'large',
      level: 2,
      basedOn: 'state',
      optionsMap: getTwoLevels(countryData),
    },
  ],
}
