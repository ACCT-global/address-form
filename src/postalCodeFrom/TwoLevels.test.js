import React from 'react'
import renderer from 'react-test-renderer'
import TwoLevels from './TwoLevels'
import { mount } from 'enzyme'
import useTwoLevels from '../country/__mocks__/useTwoLevels'
import address from '../__mocks__/newAddress'

describe('TwoLevels', () => {
  it('without first level selected', () => {
    const tree = renderer
      .create(
        <TwoLevels
          address={address}
          rules={useTwoLevels}
          onChangeAddress={jest.fn()}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('with first level selected', () => {
    const tree = renderer
      .create(
        <TwoLevels
          address={{
            ...address,
            state: 'I Región',
          }}
          rules={useTwoLevels}
          onChangeAddress={jest.fn()}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should change the first level', () => {
    const handleChange = jest.fn()
    const wrapper = mount(
      <TwoLevels
        address={address}
        rules={useTwoLevels}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'I Región' } }
    wrapper.find('select[name="state"]').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      ...address,
      state: 'I Región',
    })
  })

  it('should change the second level and postal code ', () => {
    const handleChange = jest.fn()
    const wrapper = mount(
      <TwoLevels
        address={{
          ...address,
          state: 'I Región',
        }}
        rules={useTwoLevels}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'Camiña___1150000' } }
    wrapper.find('select[name="neighborhood"]').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      ...address,
      state: 'I Región',
      neighborhood: 'Camiña',
      postalCode: '1150000',
    })
  })
})
