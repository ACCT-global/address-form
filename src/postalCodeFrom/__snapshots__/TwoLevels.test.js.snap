// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwoLevels with first level selected 1`] = `
<div>
  <label>
    region
    <select
      name="state"
      onChange={[Function]}
      value="I Región"
    >
      <option
        value=""
      >
        
      </option>
      <option
        value="I Región"
      >
        I Región
      </option>
      <option
        value="II Región"
      >
        II Región
      </option>
    </select>
  </label>
  <label>
    community
    <select
      name="neighborhood"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
      <option
        value="Camiña___1150000"
      >
        Camiña
      </option>
      <option
        value="Colchane___1160000"
      >
        Colchane
      </option>
      <option
        value="Huara___1140000"
      >
        Huara
      </option>
      <option
        value="Iquique___1100000"
      >
        Iquique
      </option>
      <option
        value="Pica___1170000"
      >
        Pica
      </option>
    </select>
  </label>
</div>
`;

exports[`TwoLevels without first level selected 1`] = `
<div>
  <label>
    region
    <select
      name="state"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        
      </option>
      <option
        value="I Región"
      >
        I Región
      </option>
      <option
        value="II Región"
      >
        II Región
      </option>
    </select>
  </label>
  <label>
    community
    <select
      name="neighborhood"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
    </select>
  </label>
</div>
`;
